OpenRA.Mods.Common/Traits/BotModules/PowerDownBotMoudle.cs

```diff
@@ -0,0 +1,148 @@
+﻿#region Copyright & License Information
+/*
+ * Copyright 2007-2019 The OpenRA Developers (see AUTHORS)
+ * This file is part of OpenRA, which is free software. It is made
+ * available to you under the terms of the GNU General Public License
+ * as published by the Free Software Foundation, either version 3 of
+ * the License, or (at your option) any later version. For more
+ * information, see COPYING.
+ */
+#endregion
+
+using System.Collections.Generic;
+using System.Linq;
+using OpenRA.Traits;
+
+namespace OpenRA.Mods.Common.Traits
+{
+	[Desc("Manages AI powerdown.")]
+	public class PowerDownBotMoudleInfo : ConditionalTraitInfo
+	{
+		[Desc("Delay (in ticks) between toggle")]
+		public readonly int Interval = 150;
+
+		[Desc("ConditionToken that powerdown")]
+		public readonly string PowerdownCondition = "powerdown";
+
+		public override object Create(ActorInitializer init) { return new PowerDownBotMoudle(init.Self, this); }
+	}
+
+	public class PowerDownBotMoudle : ConditionalTrait<PowerDownBotMoudleInfo>, IBotTick
+	{
+		readonly World world;
+		readonly Player player;
+		PowerManager playerPower;
+		int toggleTick;
+
+		public PowerDownBotMoudle(Actor self, PowerDownBotMoudleInfo info)
+			: base(info)
+		{
+			world = self.World;
+
+			if (world.Type == WorldType.Editor)
+				return;
+
+			player = self.Owner;
+		}
+
+		protected override void TraitEnabled(Actor self)
+		{
+			toggleTick = world.LocalRandom.Next(0, Info.Interval);
+			if (player == null || player.PlayerActor == null)
+			{
+				playerPower = null;
+				return;
+			}
+
+			playerPower = player.PlayerActor.TraitOrDefault<PowerManager>();
```

Comment:
This should be assigned from `Created` instead - see `HarvesterBotModule` initializing `requestUnitProduction` for an example.

Reply:
Thanks for guide. I initialize it by using the way of `BaseBuilderBotModule` but it is buggy at shellmap